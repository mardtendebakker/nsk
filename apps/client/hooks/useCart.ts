import { useEffect, useState } from 'react';
import cartStore, {
  ADD_MODULE_REQUEST,
  CLEAR_CART_REQUEST,
  Cart, EVENTS, Module, REMOVE_MODULE_REQUEST,
} from '../stores/cart';

const useCart = (): {
  state: Cart,
  removeModule:(module: Module)=> void,
  addModule:(module: Module)=> void,
  clearCart:()=> void,
  totalModulesCount: () => number
  totalAmount: () => number
} => {
  const [state, setState] = useState<Cart>(cartStore.cart);

  function cb() {
    setState(cartStore.cart);
  }

  useEffect(() => {
    EVENTS.forEach((event) => {
      cartStore.on(event, cb);
    });

    return () => {
      EVENTS.forEach((event) => {
        cartStore.removeListener(event, cb);
      });
    };
  }, []);

  return {
    state,
    removeModule: (module: Module) => {
      cartStore.emit(REMOVE_MODULE_REQUEST, module);
    },
    addModule: (module: Module) => {
      cartStore.emit(ADD_MODULE_REQUEST, module);
    },
    clearCart: () => {
      cartStore.emit(CLEAR_CART_REQUEST);
    },
    totalModulesCount: () => state.modules.length,
    totalAmount: () => {
      let total = 0;

      state.modules.forEach(({ price }) => {
        total += price;
      });

      return total;
    },
  };
};

export default useCart;
