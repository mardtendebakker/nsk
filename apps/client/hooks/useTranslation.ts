import { useEffect, useState } from 'react';
import { trans, setConfig } from 'itranslator';
import { setDefaultOptions } from 'date-fns';
import {
  nl, enUS, de, fr, ar, faIR,
} from 'date-fns/locale';
import Config from 'itranslator/lib/type/config';
import { getDefaultLocale, setDefaultLocale } from '../utils/storage';
import EventEmitter from '../utils/eventEmitter';
import nlSource from '../public/translations/nl';
import enSource from '../public/translations/en';
import deSource from '../public/translations/de';
import frSource from '../public/translations/fr';
import arSource from '../public/translations/ar';
import faSource from '../public/translations/fa';

export const LOCALE_UPDATED = 'LOCALE_UPDATED';
export type Trans = (key: string, config?: Config | undefined) => string;

export const localeMapping = {
  nl: nlSource,
  en: enSource,
  de: deSource,
  ar: arSource,
  fr: frSource,
  fa: faSource,
};

export const dateFnsMapping = {
  nl,
  en: enUS,
  de,
  fr,
  ar,
  fa: faIR,
};

class LocaleEmitter extends EventEmitter {
  #locale = getDefaultLocale() || 'nl';

  constructor() {
    super();
    setDefaultOptions({ locale: dateFnsMapping[this.#locale] });
  }

  get locale() { return this.#locale; }

  emit(event: string, payload: string) {
    if (LOCALE_UPDATED === event) {
      this.#locale = payload;

      setConfig({ source: localeMapping[payload] });
      setDefaultOptions({ locale: dateFnsMapping[payload] });
      setDefaultLocale(payload);
    }

    return super.emit(event, payload);
  }
}

export const localeStore = new LocaleEmitter();

const useTranslation = (): { updateLocale: (locale: string)=> void, trans: Trans, locale: string } => {
  const [locale, setLocale] = useState<string>(localeStore.locale);

  useEffect(() => {
    localeStore.on(LOCALE_UPDATED, setLocale);
    return () => {
      localeStore.removeListener(LOCALE_UPDATED, setLocale);
    };
  }, [setLocale]);

  const updateLocale = (newLocale: string) => {
    localeStore.emit(LOCALE_UPDATED, newLocale);
    setLocale(newLocale);
  };

  return {
    updateLocale,
    trans,
    locale,
  };
};

export default useTranslation;
