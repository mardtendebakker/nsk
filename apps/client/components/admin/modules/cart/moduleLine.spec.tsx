import { render, screen, fireEvent } from '@testing-library/react';
import ModuleLine from './moduleLine';
import useTranslation from '../../../../hooks/useTranslation';
import useCart from '../../../../hooks/useCart';

jest.mock('../../../../hooks/useTranslation', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('../../../../hooks/useCart', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('../../../confirmationDialog', () => ({
  __esModule: true,
  default: function MockConfirmationDialog({ onConfirm, onClose }) {
    return (
      <div data-testid="confirmation-dialog">
        <button onClick={onConfirm} data-testid="confirm-button" type="button">Confirm</button>
        <button onClick={onClose} data-testid="cancel-button" type="button">Cancel</button>
      </div>
    );
  },
}));

describe('ModuleLine Component', () => {
  const mockModule = {
    id: 1,
    name: 'Test Module',
    price: 99.99,
  };

  const mockTranslations = {
    removeModule: 'Remove Module',
    removeModuleConfirmation: 'Are you sure?',
    remove: 'Remove',
  };

  const mockRemoveModule = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    (useTranslation as jest.Mock).mockReturnValue({
      trans: (key: string) => mockTranslations[key],
    });

    (useCart as jest.Mock).mockReturnValue({
      removeModule: mockRemoveModule,
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<ModuleLine module={mockModule} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders module information correctly', () => {
    render(<ModuleLine module={mockModule} />);

    expect(screen.getByText(mockModule.name)).toBeInTheDocument();
    expect(screen.getByText('â‚¬ 99,99')).toBeInTheDocument();
    expect(screen.getByText(mockTranslations.remove)).toBeInTheDocument();
  });

  it('shows confirmation dialog when remove is clicked', () => {
    render(<ModuleLine module={mockModule} />);

    fireEvent.click(screen.getByText(mockTranslations.remove));
    expect(screen.getByTestId('confirmation-dialog')).toBeInTheDocument();
  });

  it('closes dialog without removing module when cancelled', () => {
    render(<ModuleLine module={mockModule} />);

    fireEvent.click(screen.getByText(mockTranslations.remove));
    fireEvent.click(screen.getByTestId('cancel-button'));

    expect(mockRemoveModule).not.toHaveBeenCalled();
  });
});
