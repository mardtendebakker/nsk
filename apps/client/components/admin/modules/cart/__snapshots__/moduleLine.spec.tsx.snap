// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleLine Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-17sib13"
  >
    <div
      data-testid="confirmation-dialog"
    >
      <button
        data-testid="confirm-button"
        type="button"
      >
        Confirm
      </button>
      <button
        data-testid="cancel-button"
        type="button"
      >
        Cancel
      </button>
    </div>
    <div
      class="MuiBox-root css-1lekzkb"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
      >
        Test Module
      </h2>
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        € 99,99
      </p>
    </div>
    <div
      class="MuiBox-root css-1bvc4cc"
    >
      <span
        class="MuiTypography-root MuiTypography-caption css-n9mh9m-MuiTypography-root"
      >
        Remove
      </span>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-1x67s69-MuiDivider-root"
    />
  </div>
</DocumentFragment>
`;
